onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -509,68 +602,175 @@
     size_t BlockStrideQuantB
 )
 {
-    auto impl0_reference = [&]() {
-        constexpr size_t BlkBitWidth = 4;
-        constexpr size_t SubBlkLen = 16;
+    constexpr size_t BlkBitWidth = 4;
 
-        float* Dst = FpData;
+    float* Dst = FpData;
 
-        const std::byte* QuantBDataCol = QuantBData;
-        const float* QuantBScaleCol = QuantBScale;
-        const std::byte* QuantBZeroPointCol = QuantBZeroPoint;
+    const std::byte* QuantBDataCol = QuantBData;
+    const float* QuantBScaleCol = QuantBScale;
+    [[maybe_unused]] const std::byte* QuantBZeroPointCol = QuantBZeroPoint;  // only used if HasZeroPoint is true
 
-        for (size_t n = 0; n < CountN; n += 16) {
-            const size_t nnlen = std::min(CountN - n, size_t{16});
+    const size_t StrideQuantBData = BlockStrideQuantB * MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+    [[maybe_unused]] const size_t StrideQuantBZeroPoint =  // only used if HasZeroPoint is true
+        MlasQNBitZeroPointsForBlksSizeInBytes<BlkBitWidth>(BlockStrideQuantB);
 
-            for (size_t nn = 0; nn < nnlen; ++nn) {
-                for (size_t k = 0, k_blk_idx = 0; k < CountK; k += BlkLen, k_blk_idx += 1) {
-                    const size_t kklen = std::min(CountK - k, BlkLen);
+    //
+    // Proceed down 16 column-wide regions of B. Dequantize and write output 16 x 16 elements at a time.
```

Comment:
>16 [](http://example.com/codeflow?start=20&length=2)

did you try 32 column? ARM can load 32 items with one instruction

Reply:
if it can get better perf, it would be better to specialize the code base on block size

Reply:
with 32 rows and 4 columns, it requires 8 * 4 float32x4_t vector registers for the result. so we'd already be at the 32 register limit not accounting for the intermediate registers used in the conversion.

it's also possible to do less than 4 columns at once, but I found that writing the (transposed) result from 4 columns with one `vst1q_f32` has better performance (added this in dbcdd232b9).